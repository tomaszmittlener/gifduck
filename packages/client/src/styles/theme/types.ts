import { BuildContainerGrid, BuildSpanGrid } from '../grid'

export interface ThemeColorsPrimary {
  default: string
  defaultInverted: string
  secondary: string
  third: string
  [key: string]: string
}

export interface ThemeColorsAccent {
  default: string
  defaultInverted: string
  [key: string]: string
}

export interface ThemeColorsGrays {
  default: string
  white: string
  black: string
  [key: string]: string
}

export interface ThemeColorsStatus {
  success: string
  error: string
  [key: string]: string
}

export interface ThemeColors {
  primary: Readonly<ThemeColorsPrimary>
  accent: Readonly<ThemeColorsAccent>
  grays: Readonly<ThemeColorsGrays>
  status: Readonly<ThemeColorsStatus>
}

export interface ThemeTypographyFonts {
  default: string
  secondary: string
}

export interface ThemeTypographyWeights {
  regular: number
  bold: number
}
export interface ThemeTypography {
  fonts: Readonly<ThemeTypographyFonts>
  weights: Readonly<ThemeTypographyWeights>
}

export interface ThemeGrid {
  container: BuildContainerGrid
  span: BuildSpanGrid
}
export interface ThemeBreakpoints {
  tablet: string
  desktop: string
}

export interface Theme {
  colors: Readonly<ThemeColors>
  typography: Readonly<ThemeTypography>
  ms: (step: number) => string
  layers: Readonly<ThemeLayers>
  grid: Readonly<ThemeGrid>
  breakpoints: ThemeBreakpoints
}

export interface ThemeLayers {
  bottom: number
  middleBottom: number
  middle: number
  middleTop: number
  top: number
  [key: string]: number
}
