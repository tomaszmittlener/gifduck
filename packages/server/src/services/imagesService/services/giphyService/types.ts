/**
 *
 * Giphy search Api
 *
 */

/**
 * Giphy Search Endpoint query params.
 * TODO: Add rest of available options
 * */
export type GiphySearchQuery = {
  /** Search query term or phrase. */
  q?: string
  /** GIPHY API Key. */
  api_key: string
  /** The maximum number of objects to return. (Default: “25”) */
  limit?: number
  /** Specifies the starting position of the results. Defaults to 0. */
  offset?: number
}

/** Giphy object with meta data */
export interface GiphyMetaObject {
  msg: string
  status: number
  response_id: string
}

/** Giphy object with pagination data */
export interface GiphyPaginationObject {
  offset: number
  total_count: number
  count: number
}

/** Giphy object with available images data (incl. height, width, format etc.) */
export interface GiphyImagesObject {
  fixed_height: {
    url: string
    width: string
    height: string
    size: string
    mp4: string
    mp4_size: string
    webp: string
    webp_size: string
  }
  fixed_height_still: {
    url: string
    width: string
    height: string
  }
  fixed_height_downsampled: {
    url: string
    width: string
    height: string
    size: string
    webp: string
    webp_size: string
  }
  fixed_width: {
    url: string
    width: string
    height: string
    size: string
    mp4: string
    mp4_size: string
    webp: string
    webp_size: string
  }
  fixed_width_still: {
    url: string
    width: string
    height: string
  }
  fixed_width_downsampled: {
    url: string
    width: string
    height: string
    size: string
    webp: string
    webp_size: string
  }
  fixed_height_small: {
    url: string
    width: string
    height: string
    size: string
    webp: string
    webp_size: string
  }
  fixed_height_small_still: {
    url: string
    width: string
    height: string
  }
  fixed_width_small: {
    url: string
    width: string
    height: string
    size: string
    webp: string
    webp_size: string
  }
  fixed_width_small_still: {
    url: string
    width: string
    height: string
  }
  downsized: {
    url: string
    width: string
    height: string
    size: string
  }
  downsized_still: {
    url: string
    width: string
    height: string
  }
  downsized_large: {
    url: string
    width: string
    height: string
    size: string
  }
  original: {
    url: string
    width: string
    height: string
    size: string
    frames: string
    mp4: string
    mp4_size: string
    webp: string
    webp_size: string
  }
  original_still: {
    url: string
    width: string
    height: string
  }
}

/** Giphy base image object */
export interface GiphyGIFObject {
  type: string
  id: string
  slug: string
  url: string
  bitly_gif_url: string
  bitly_url: string
  embed_url: string
  username: string
  source: string
  rating: string
  caption: string
  content_url: string
  source_tld: string
  source_post_url: string
  import_datetime: string
  trending_datetime: string
  images: GiphyImagesObject
  meta: GiphyMetaObject
}

/** Response for giphy image search request */
export interface GiphyImageResponse {
  data: GiphyGIFObject[]
  pagination: GiphyPaginationObject
  meta: GiphyMetaObject
}
