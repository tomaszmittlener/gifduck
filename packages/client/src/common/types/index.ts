export enum FetchStatus {
  Loading = 'loading',
  Loaded = 'loaded',
  Error = 'error',
}
