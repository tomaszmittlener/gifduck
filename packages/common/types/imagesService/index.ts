export interface ImageObject {
  height: string
  width: string
  url: string
}

export interface ImageData {
  preview: ImageObject
  image: ImageObject
  id: string
}

export interface ImagesResponse {
  data: ImageData[]
}

export enum SearchQueryKeys {
  searchText = 'q',
}

export interface SearchQuery {
  [SearchQueryKeys.searchText]: string
}
