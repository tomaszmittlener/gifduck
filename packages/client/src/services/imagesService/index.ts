import Axios, { AxiosPromise } from 'axios'
import { SearchQuery, ImagesSearchResponse } from '@gifduck/types/imagesService'
import { stringifyQuery } from 'common/queryString'

const BASE_URL = process.env.IMAGES_SEARCH_BAES_URL

const instance = Axios.create({
  baseURL: BASE_URL + '/api/v1/images',
})

interface ImagesService {
  getImages: (searchQuery: SearchQuery) => AxiosPromise<ImagesSearchResponse>
}

const imagesService: ImagesService = {
  getImages: async (query: SearchQuery) => {
    return instance.get(stringifyQuery(query))
  },
}

export default imagesService
