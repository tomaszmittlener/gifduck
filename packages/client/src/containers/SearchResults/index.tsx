import React, { FunctionComponent } from 'react'
import styled from 'styled-components'
import { ImageData } from '@gifduck/common-types/imagesService'

import Gallery from 'components/Gallery'
import GalleryItem from 'components/GalleryItem'
import Image from 'components/Image'

interface SearchResultsProps {
  className?: string
  testId?: string
  results: ImageData[]
}

const Container = styled.section`
  background-color: ${({ theme }) => theme.colors.grays.black};
  ${({ theme }) => theme.grid.container()};
`

const StyledGallery = styled(Gallery)`
  ${({ theme }) => theme.grid.span(1, 8)};
`

const SearchResults: FunctionComponent<SearchResultsProps> = ({ className, testId, results }) => {
  return (
    <Container data-testid={testId} className={className}>
      <StyledGallery>
        {results.map(result => (
          <GalleryItem key={result.id}>
            <Image
              src={result.image.url}
              width={result.image.width}
              height={result.image.height}
              aspectRatio={result.image.aspectRatio}
              preSrc={result.preview.url}
            />
          </GalleryItem>
        ))}
      </StyledGallery>
    </Container>
  )
}

export default SearchResults
