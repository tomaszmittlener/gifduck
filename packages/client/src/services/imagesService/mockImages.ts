import { ImagesResponse } from '@gifduck/common-types/imagesService'

const mockImages: ImagesResponse = {
  data: [
    {
      id: '1',
      image: {
        height: '500',
        url:
          'https://media2.giphy.com/media/Cmr1OMJ2FN0B2/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '500',
      },
      preview: {
        width: '500',
        url:
          'https://media2.giphy.com/media/Cmr1OMJ2FN0B2/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '500',
      },
    },
    {
      id: '2',
      image: {
        height: '480',
        url:
          'https://media0.giphy.com/media/65OOoKUgwJlJpRma8O/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '480',
      },
      preview: {
        width: '248',
        url:
          'https://media0.giphy.com/media/65OOoKUgwJlJpRma8O/giphy-downsized_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy-downsized_s.gif',
        height: '248',
      },
    },
    {
      id: '3',
      image: {
        height: '332',
        url:
          'https://media2.giphy.com/media/vFKqnCdLPNOKc/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '444',
      },
      preview: {
        width: '444',
        url:
          'https://media2.giphy.com/media/vFKqnCdLPNOKc/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '332',
      },
    },
    {
      id: '4',
      image: {
        height: '353',
        url:
          'https://media0.giphy.com/media/BVStb13YiR5Qs/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '500',
      },
      preview: {
        width: '500',
        url:
          'https://media0.giphy.com/media/BVStb13YiR5Qs/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '353',
      },
    },
    {
      id: '5',
      image: {
        height: '358',
        url:
          'https://media3.giphy.com/media/3otPongOYeG9iINM8o/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '452',
      },
      preview: {
        width: '452',
        url:
          'https://media3.giphy.com/media/3otPongOYeG9iINM8o/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '358',
      },
    },
    {
      id: '6',
      image: {
        height: '260',
        url:
          'https://media2.giphy.com/media/icUEIrjnUuFCWDxFpU/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '480',
      },
      preview: {
        width: '480',
        url:
          'https://media2.giphy.com/media/icUEIrjnUuFCWDxFpU/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '260',
      },
    },
    {
      id: '7',
      image: {
        height: '184',
        url:
          'https://media3.giphy.com/media/IThjAlJnD9WNO/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '245',
      },
      preview: {
        width: '245',
        url:
          'https://media3.giphy.com/media/IThjAlJnD9WNO/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '184',
      },
    },
    {
      id: '8',
      image: {
        height: '722',
        url:
          'https://media2.giphy.com/media/YrZECW1GgBkqat6F0B/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '800',
      },
      preview: {
        width: '800',
        url:
          'https://media2.giphy.com/media/YrZECW1GgBkqat6F0B/giphy_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy_s.gif',
        height: '722',
      },
    },
    {
      id: '9',
      image: {
        height: '480',
        url:
          'https://media1.giphy.com/media/MWSRkVoNaC30A/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '384',
      },
      preview: {
        width: '200',
        url:
          'https://media1.giphy.com/media/MWSRkVoNaC30A/giphy-downsized_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy-downsized_s.gif',
        height: '250',
      },
    },
    {
      id: '10',
      image: {
        height: '270',
        url:
          'https://media2.giphy.com/media/3oKIPsx2VAYAgEHC12/giphy.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy.gif',
        width: '480',
      },
      preview: {
        width: '384',
        url:
          'https://media2.giphy.com/media/3oKIPsx2VAYAgEHC12/giphy-downsized_s.gif?cid=e4e1ee367ecd732a2a49e8edbd6c36992b2bed00708ee5bd&rid=giphy-downsized_s.gif',
        height: '216',
      },
    },
    {
      id: '11',
      image: {
        height: '360',
        url: 'https://pixabay.com/get/5ee3dd4a4253b108f5d0846096213f7d133cdbe74e507441752f7cd09f44c5_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2015/07/10/09/31/happy-838981_150.jpg',
        height: '84',
      },
    },
    {
      id: '12',
      image: {
        height: '640',
        url: 'https://pixabay.com/get/57e0d7444850a514f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2015/11/06/12/43/mask-1027229_150.jpg',
        height: '150',
      },
    },
    {
      id: '13',
      image: {
        height: '360',
        url: 'https://pixabay.com/get/57e3d144425aaa14f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2016/04/23/17/06/seal-1347886_150.jpg',
        height: '84',
      },
    },
    {
      id: '14',
      image: {
        height: '544',
        url: 'https://pixabay.com/get/57e7d2404c53a414f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.png',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2016/10/27/07/53/bird-1773618_150.png',
        height: '127',
      },
    },
    {
      id: '15',
      image: {
        height: '640',
        url: 'https://pixabay.com/get/52e7d74b4e50b108f5d0846096213f7d133cdbe74e507441752f7cd09f44c5_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2014/10/04/11/45/cow-472842_150.jpg',
        height: '150',
      },
    },
    {
      id: '16',
      image: {
        height: '274',
        url: 'https://pixabay.com/get/54e0d4464e54aa14f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2017/01/28/14/19/hello-2015466_150.jpg',
        height: '64',
      },
    },
    {
      id: '17',
      image: {
        height: '640',
        url: 'https://pixabay.com/get/57e2dc414355ae14f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.png',
        width: '445',
      },
      preview: {
        width: '104',
        url: 'https://cdn.pixabay.com/photo/2016/03/31/15/04/alien-1292972_150.png',
        height: '150',
      },
    },
    {
      id: '18',
      image: {
        height: '360',
        url: 'https://pixabay.com/get/55e7d2444a51ab14f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2018/10/27/17/29/window-3777037_150.jpg',
        height: '84',
      },
    },
    {
      id: '19',
      image: {
        height: '424',
        url: 'https://pixabay.com/get/5ee7d5404f50b108f5d0846096213f7d133cdbe74e507441752f7cd09f44c5_640.jpg',
        width: '640',
      },
      preview: {
        width: '150',
        url: 'https://cdn.pixabay.com/photo/2015/08/01/15/19/woman-watching-870352_150.jpg',
        height: '99',
      },
    },
    {
      id: '20',
      image: {
        height: '640',
        url: 'https://pixabay.com/get/57e8dc4a4b52a514f6da8c7dda793e761639dde0514c704c7d2f7bd39345cc58_640.jpg',
        width: '453',
      },
      preview: {
        width: '106',
        url: 'https://cdn.pixabay.com/photo/2016/12/11/10/22/businessman-1899109_150.jpg',
        height: '150',
      },
    },
  ],
}

export default mockImages
