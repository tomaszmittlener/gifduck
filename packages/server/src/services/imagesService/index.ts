import { ImageData, SearchQuery } from '@gifduck/types/imagesService'
import giphyService from './utilities/giphyService'
import pixabayService from './utilities/pixabayService'
import formatGiphyResponse from './utilities/formatGiphyResponse'
import formatPixabayResponse from './utilities/formatPixabayResponse'

const imagesService = {
  search: async (searchQuery: SearchQuery): Promise<ImageData[]> => {
    try {
      const giphyData = await giphyService.search(searchQuery)
      const giphyImages = formatGiphyResponse(giphyData.data)
      const pixabayData = await pixabayService.search(searchQuery)
      const pixabayImages = await formatPixabayResponse(pixabayData.data)

      return [...giphyImages, ...pixabayImages]
    } catch (e) {
      throw e
    }
  },
}

export default imagesService
