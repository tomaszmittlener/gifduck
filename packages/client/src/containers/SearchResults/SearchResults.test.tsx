import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { ImageData } from '@gifduck/types/imagesService'
import { MemoryRouter } from 'react-router-dom'
import Theme from 'styles/theme'

import SearchResults from './'

const results: ImageData[] = [
  {
    id: '1',
    image: {
      height: 100,
      url: 'image-url',
      width: 200,
      aspectRatio: 1.2222,
    },
    preview: {
      width: 50,
      url: 'preview-url',
      height: 100,
      aspectRatio: 0.44444,
    },
  },
]

const defaultProps = {
  className: 'class-name',
  testId: 'test-id',
  results,
}

afterEach(cleanup)

describe('SearchResults', () => {
  test('renders correctly', async () => {
    const { container } = render(
      <ThemeProvider theme={Theme}>
        <MemoryRouter>
          <SearchResults {...defaultProps} />
        </MemoryRouter>
      </ThemeProvider>,
    )
    expect(container).toMatchSnapshot()
  })

  test('passes className to container', async () => {
    const { getByTestId } = render(
      <ThemeProvider theme={Theme}>
        <MemoryRouter>
          <SearchResults {...defaultProps} />
        </MemoryRouter>
      </ThemeProvider>,
    )

    expect(getByTestId(defaultProps.testId)).toHaveClass(defaultProps.className)
  })
})
