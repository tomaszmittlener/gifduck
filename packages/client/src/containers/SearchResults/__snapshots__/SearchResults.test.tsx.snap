// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults renders correctly 1`] = `
.c1 {
  padding: 0;
  display: grid;
  grid-row-gap: 0.8888888888888888rem;
  grid-column-gap: 1%;
  grid-template-columns: 1fr;
}

.c3 {
  padding: 0;
  margin: 0;
  list-style: none;
  grid-area: auto / 1 / span 2 / 5;
}

.c4 {
  margin: 0;
  position: relative;
  width: 100%;
  height: 100%;
}

.c4:before {
  display: block;
  content: '';
  width: 100%;
  padding-top: 122.22%;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 1;
  z-index: 1100;
}

.c0 {
  padding: 0 1rem;
  grid-column-gap: 0.49327018427257213rem;
  display: grid;
  grid-template-columns: repeat(8,1fr);
  width: 100%;
  max-width: 1440px;
  height: 100%;
}

.c2 {
  grid-column: 1 / 9;
}

@media screen and (min-width:768px) {
  .c1 {
    grid-row-gap: 1.601806640625rem;
    grid-template-columns: 6% 14% 11% 15% 5% 10% 9% 23%;
  }
}

@media screen and (min-width:768px) {
  .c3:first-of-type {
    grid-area: auto / 3 / span 1 / 7;
  }

  .c3:nth-of-type(2n) {
    grid-area: auto / 4 / span 1 / 9;
  }

  .c3:nth-of-type(3n) {
    grid-area: auto / 2 / span 1 / 7;
  }

  .c3:nth-of-type(4n) {
    grid-area: auto / 5 / span 1 / 9;
  }

  .c3:nth-of-type(7n) {
    grid-area: auto / 4 / span 1 / 6;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    padding: 0 1.423828125rem;
  }
}

@media screen and (min-width:1024px) {
  .c0 {
    padding: 0 2.0272865295410156rem;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    grid-column-gap: 1rem;
  }
}

<div>
  <section
    class="c0 class-name"
    data-testid="test-id"
  >
    <ul
      class="c1 c2"
    >
      <li
        class="c3"
      >
        <figure
          class="c4"
        >
          <img
            class="c5"
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
          />
        </figure>
      </li>
    </ul>
  </section>
</div>
`;
